iceoryx_posh/test/moduletests/test_popo_subscriber_port.cpp

```diff
@@ -301,3 +301,165 @@ TEST_F(SubscriberPort_test, nackWhenNotWaitingForResultsInError)
+    EXPECT_THAT(caproMessageResponse.m_historyCapacity, Eq(0u));
+}
+
+TEST_F(SubscriberPortMultiProducer_test, unsubscribeInWaitForOfferResultsInNotSubscribed)
+{
+    m_sutUserSideMultiProducer.subscribe();
+    m_sutRouDiSideMultiProducer.getCaProMessage(); // only RouDi changes state
+    m_sutUserSideMultiProducer.unsubscribe();
+    m_sutRouDiSideMultiProducer.getCaProMessage(); // only RouDi changes state
+
+    const auto subscriptionState = m_sutUserSideMultiProducer.getSubscriptionState();
+
+    EXPECT_THAT(subscriptionState, Eq(iox::SubscribeState::NOT_SUBSCRIBED));
+}
+
+TEST_F(SubscriberPortMultiProducer_test, StopOfferInWaitForOfferResultsInWaitForOffer)
+{
+    m_sutUserSideMultiProducer.subscribe();
+    m_sutRouDiSideMultiProducer.getCaProMessage(); // only RouDi changes state
+    iox::capro::CaproMessage caproMessage(iox::capro::CaproMessageType::ACK,
+                                          iox::capro::ServiceDescription("x", "y", "z"));
+    m_sutRouDiSideMultiProducer.dispatchCaProMessage(caproMessage);
+    caproMessage.m_type = iox::capro::CaproMessageType::STOP_OFFER;
+    m_sutRouDiSideMultiProducer.dispatchCaProMessage(caproMessage);
+
+    const auto subscriptionState = m_sutUserSideMultiProducer.getSubscriptionState();
+
+    EXPECT_THAT(subscriptionState, Eq(iox::SubscribeState::WAIT_FOR_OFFER));
+}
+
+TEST_F(SubscriberPortMultiProducer_test, unsubscribeInWaitForOfferTriggersUnsubMessage)
+{
+    m_sutUserSideMultiProducer.subscribe();
+    m_sutRouDiSideMultiProducer.getCaProMessage(); // only RouDi changes state
+    iox::capro::CaproMessage caproMessage(iox::capro::CaproMessageType::ACK,
+                                          iox::capro::ServiceDescription("x", "y", "z"));
+    m_sutRouDiSideMultiProducer.dispatchCaProMessage(caproMessage);
+    m_sutUserSideMultiProducer.unsubscribe();
+
+    auto maybeCaproMessage = m_sutRouDiSideMultiProducer.getCaProMessage();
+
+    EXPECT_TRUE(maybeCaproMessage.has_value());
+    auto caproMessageResponse = maybeCaproMessage.value();
+    EXPECT_THAT(caproMessageResponse.m_type, Eq(iox::capro::CaproMessageType::UNSUB));
+    EXPECT_THAT(caproMessageResponse.m_serviceDescription, Eq(iox::capro::ServiceDescription("x", "y", "z")));
+}
+
+TEST_F(SubscriberPortMultiProducer_test, invalidMessageResultsInError)
+{
+    auto errorHandlerCalled{false};
+    auto errorHandlerGuard = iox::ErrorHandler::SetTemporaryErrorHandler([&errorHandlerCalled](
+        const iox::Error, const std::function<void()>, const iox::ErrorLevel) { errorHandlerCalled = true; });
+    iox::capro::CaproMessage caproMessage(iox::capro::CaproMessageType::UNSUB,
+                                          iox::capro::ServiceDescription("x", "y", "z"));
+
+    auto maybeCaproMessage = m_sutRouDiSideMultiProducer.dispatchCaProMessage(caproMessage);
+
+    EXPECT_FALSE(maybeCaproMessage.has_value());
+    EXPECT_TRUE(errorHandlerCalled);
```

Comment:
It should be also against the corresponding errortype tested:
This can be done like this:

    auto errorHandlerCalled{false};
    iox::Error receivedError;

    auto errorHandlerGuard = iox::ErrorHandler::SetTemporaryErrorHandler(
        [&errorHandlerCalled,
         &receivedError](const iox::Error error, const std::function<void()>, const iox::ErrorLevel) {
            errorHandlerCalled = true;
            receivedError = error;
        });

    ASSERT_TRUE(errorHandlerCalled);
    ASSERT_THAT(receivedError, Eq(iox::Error::kPOPO__CAPRO_PROTOCOL_ERROR));